miden/src/repl/mod.rs

```diff
@@ -0,0 +1,311 @@
+/// If memory at the specified address has been initialized:
+/// `!mem[9]`
+/// >> 9: [9, 10, 2, 0]
+///
+/// If memory at the specified address has not been initialized:
+/// `!mem[8]`
+/// >> Memory at address 8 is empty
+
+/// Initiates the Miden Repl tool.
+pub fn start_repl() {
+    let mut program_lines: Vec<String> = Vec::new();
+    let _functions_miden = "".to_string();
+
+    println!("========================== Miden REPL ============================");
+    println!();
+    // prints out all the available commands in the Miden Repl tool.
+    print_instructions();
+
+    let mut stack_print_flag = false;
+
+    // state of the entire memory at the latest clock cycle.
+    let mut memory: Vec<(u64, Word)> = Vec::new();
+
+    // initialising readline.
+    let mut rl = Editor::<()>::new().unwrap();
+    loop {
+        let program = format!(
+            "begin\n{}\nend",
+            program_lines
+                .iter()
+                .map(|l| format!("    {}", l))
+                .collect::<Vec<_>>()
+                .join("\n")
+        );
+
+        let result = execute(program.clone());
+
+        let mut result_string = "0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0".to_string();
+        if !program_lines.is_empty() {
+            match result {
+                Ok((mem, stack_state)) => {
+                    result_string = print_stack(stack_state);
+                    if stack_print_flag {
+                        println!("{}", result_string);
+                    }
+                    memory = mem;
+                }
+                Err(e) => {
+                    println!("{}", format!("Error running program: {:?}", e));
+                    program_lines.pop();
+                }
+            }
+        }
+        let readline = rl.readline(">> ");
+        match readline {
+            Ok(line) => {
+                if line == "!program" {
+                    println!("{}", program);
+                    stack_print_flag = false;
+                } else if line == "!help" {
+                    // prints out all the available commands in the Miden Repl tool.
+                    print_instructions();
+                    stack_print_flag = false;
+                } else if line == "!mem" {
+                    for (addr, mem) in &memory {
+                        // each memory register converted from Felt to an integer.
+                        let mem_int = print_mem_address(mem);
+                        println!("{}: {:?}", addr, mem_int)
+                    }
+                    stack_print_flag = false;
+                } else if line.len() > 6 && &line[..5] == "!mem[" {
+                    // if user wants to see the state of a particular address in a memory, the input should be atleast
+                    // of length 5.
+
+                    // extracting the digits from the input text given by the user.
+                    let result = read_mem_address(&line);
```

Comment:
The name `result` is not meaningful here - I would have named it something like `addr`. But also, we could combine this line with the `match` statement later as:
```Rust
match read_mem_address(&line) {
    Ok(addr) => { }
    Err(msg) => println!("{}", msg),
}
```